src/webgpu/api/validation/buffer_mapping.spec.ts

```diff
@@ -0,0 +1,763 @@
+      await p;
+      this.shouldResolve(p);
+    } else {
+      this.expectValidationError(async () => {
+        const p = buffer.mapAsync(mode, offset, size);
+        await p;
+        this.shouldReject(rejectName!, p);
+      });
+    }
+  }
+
+  async testGetMappedRangeCall(
+    success: boolean,
+    buffer: GPUBuffer,
+    offset?: number,
+    size?: number
+  ) {
+    if (success) {
+      const data = buffer.getMappedRange(offset, size);
+      this.expect(data instanceof ArrayBuffer);
+      if (size !== undefined) {
+        this.expect(data.byteLength === size);
+      }
+    } else {
+      this.shouldThrow('OperationError', () => {
+        buffer.getMappedRange(offset, size);
+      });
+    }
+  }
+
+  createMappableBuffer(type: GPUMapModeFlags, size: number): GPUBuffer {
+    switch (type) {
+      case GPUMapMode.READ:
+        return this.device.createBuffer({
+          size,
+          usage: GPUBufferUsage.MAP_READ,
+        });
+      case GPUMapMode.WRITE:
+        return this.device.createBuffer({
+          size,
+          usage: GPUBufferUsage.MAP_WRITE,
+        });
+      default:
+        unreachable();
+    }
+  }
+}
+
+export const g = makeTestGroup(F);
+
+const kMapModeOptions = poptions('mapMode', [GPUMapMode.READ, GPUMapMode.WRITE]);
+const kOffsetAlignment = 8;
+const kSizeAlignement = 4;
+
+g.test('mapAsync_usage')
+  .desc(
+    `Test the usage validation for mapAsync.
+
+  For each buffer usage:
+  For GPUMapMode.READ, GPUMapMode.WRITE, and 0:
+    Test that the mapAsync call is valid iff the mapping usage is not 0 and the buffer usage
+    the mapMode flag.`
+  )
+  .params(
+    params()
+      .combine([
+        { mapMode: GPUMapMode.READ, validUsage: GPUBufferUsage.MAP_READ },
+        { mapMode: GPUMapMode.WRITE, validUsage: GPUBufferUsage.MAP_WRITE },
+        // Using mapMode 0 is never valid, so there is no validUsage.
+        { mapMode: 0, validUsage: null },
+      ])
+      .combine(poptions('usage', kBufferUsages))
+  )
+  .fn(async t => {
+    const { mapMode, validUsage, usage } = t.params;
+
+    const buffer = t.device.createBuffer({
+      size: 16,
+      usage,
+    });
+
+    const success = usage === validUsage;
+    await t.testMapAsyncCall(success, 'OperationError', buffer, mapMode);
+  });
+
+g.test('mapAsync_invalidBuffer')
+  .desc('Test that mapAsync is an error when called on an invalid buffer.')
+  .params(kMapModeOptions)
+  .fn(async t => {
+    const { mapMode } = t.params;
+    const buffer = t.getErrorBuffer();
+    await t.testMapAsyncCall(false, 'OperationError', buffer, mapMode);
+  });
+
+g.test('mapAsync_state_destroyed')
+  .desc('Test that mapAsync is an error when called on a destroyed buffer.')
+  .params(kMapModeOptions)
+  .fn(async t => {
+    const { mapMode } = t.params;
+    const buffer = t.createMappableBuffer(mapMode, 16);
+    buffer.destroy();
+    await t.testMapAsyncCall(false, 'OperationError', buffer, mapMode);
+  });
```

Comment:
I think it makes sense for other tests because the usage is only used server-side, but mapping usage could be checked client-side so we need to make sure the usage flags are correct (which a helper wouldn't give us). WDYT?

Reply:
createBufferWithState does take a descriptor. It ignores it if you ask for an error texture, but it doesn't necessarily have to.

Reply:
I had some trouble refactoring to use the new helper because all tests are parametrized with the `mapAsync` mode and not the buffer usage so the helper requires extra steps to use (which kinda defeat the purpose).

I still included it in in the latest patchset.